src/lib/generateUserProviderConfig.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { color } from "@oclif/color";
+
+import { isInteractive } from "./commandObj.js";
+import type { UserProvidedConfig, Offer } from "./configs/project/provider.js";
+import {
+  defaultNumberProperties,
+  type NumberProperty,
+  numberProperties,
+} from "./const.js";
+import { commaSepStrToArr } from "./helpers/utils.js";
+import { validatePositiveNumberOrEmpty } from "./helpers/validations.js";
+import { checkboxes, confirm, input } from "./prompt.js";
+
+async function promptToSetNumberProperty(
+  offer: Offer,
+  property: NumberProperty,
+) {
+  const propertyStr = await input({
+    message: `Enter ${color.yellow(property)}`,
+    validate: validatePositiveNumberOrEmpty,
+    default: `${defaultNumberProperties[property]}`,
+  });
+
+  offer[property] = Number(propertyStr);
+}
+
+const DEFAULT_NUMBER_OF_NOXES = 3;
+
+export type ProviderConfigArgs = {
+  numberOfNoxes?: number | undefined;
+};
+
+export async function generateUserProviderConfig({
```

Comment:
I suggest to refactor body of this function. See above.
This refactoring will make this function cleaner, moving implementation details in more specific functions which facilitate a level of abstraction.

If you agree please apply this refactoring to all `do... while` constructions

Reply:
sounds ok